File: python/forth/forth.py
import inspect
from itertools import chain
import string


# Define dictionary OPS that maps operators to corresponding lambda functions
OPS = {
    '+':    lambda x, y: [y + x],   # Addition
    '-':    lambda x, y: [y - x],   # Subtraction
    '*':    lambda x, y: [y * x],   # Multiplication
    '/':    lambda x, y: [y // x],  # Integer Division
    'dup':  lambda x:    [x, x],    # Duplicate the top item on the stack
    'drop': lambda _:    [],        # Remove the top item from the stack
    'swap': lambda x, y: [x, y],    # Swap the top two items on the stack
    'over': lambda x, y: [y, x, y]  # Copy the second item on the stack to the top
}

# Custom Exception classes for error handling
class StackUnderflowError(Exception):
    def __init__(self):
        super().__init__('Insufficient number of items in stack')

class ZeroDivisonError(Exception):
    def __init__(self):
        super().__init__('divide by zero')

# Function to check if an element is a valid number
def is_number(elem):
    return elem and (set(elem) < set(string.digits)
                     or (elem[0] == '-' and is_number(elem[1:])))

# Function to apply an element to the stack
def apply(stack, elem):
    if is_number(elem):
        stack.append(int(elem))
    elif elem in OPS:
        op = OPS[elem]
        count = len(inspect.signature(op).parameters)
        if len(stack) < count:
            raise StackUnderflowError
        stack.extend(op(*(stack.pop() for x in range(count))))
    else:
        raise ValueError('undefined operation')

# Function to substitute custom definitions in input data
def substitute(custom, elems):
    return list(chain(*(custom[x] if x in custom else [x] for x in elems)))

# Main evaluation function
def evaluate(input_data):
    stack = []
    custom = {}
    try:
        for line in input_data:
            elems = line.lower().split()
            if elems[0] == ':':
                assert elems[-1] == ';'
                op = elems[1]
                if is_number(op):
                    raise ValueError('illegal operation')
                custom[op] = substitute(custom, elems[2:-1])
            else:
                elems = substitute(custom, elems)
                for elem in elems:
                    apply(stack, elem)
        return stack
    except ZeroDivisionError:
        raise ZeroDivisionError('divide by zero')

File: javascript/forth/forth.js
const OPERATION_MAP = {
  "+": (a, b) => a + b,
  "-": (a, b) => a - b,
  "*": (a, b) => a * b,
  "/": (a, b) => {
    if (b === 0) throw Error("Division by zero");
    return Math.floor(a / b);
  },
};
const OPERATORS = new Set(Object.keys(OPERATION_MAP));
const getVars = (stack, count = 1) => {
  if (stack.length < count) throw Error("Stack empty");
  return stack.splice(count * -1);
};
const COMMAND_MAP = {
  dup: (stack) => {
    const [a] = getVars(stack);
    stack.push(a);
    stack.push(a);
  },
  drop: (stack) => getVars(stack),
  swap: (stack) => {
    const [a, b] = getVars(stack, 2);
    stack.push(b, a);
  },
  over: (stack) => {
    const [a, b] = getVars(stack, 2);
    stack.push(a, b, a);
  },
};
const STACK_COMMANDS = new Set(Object.keys(COMMAND_MAP));

export class Forth {
  constructor() {
    this._stack = [];
    this._dictionary = {};
  }

  _defineCommand(terms) {
    const command = terms[1];

    if (command.match(/^-?\d+$/)) throw Error("Invalid definition");

    const action = this._expandActions(terms.slice(2, -1)).join(" ");

    this._dictionary[command] = action;
  }

  _expandActions(terms) {
    return terms.flatMap((term) => this._dictionary[term]?.split(" ") ?? term);
  }

  _performArithmetic(operator) {
    const b = this._stack.pop();
    const a = this._stack.pop();

    if (a === undefined || b === undefined) throw Error("Stack empty");

    const result = OPERATION_MAP[operator](a, b);
    this._stack.push(result);
  }

  _manipulateStack(command) {
    COMMAND_MAP[command](this._stack);
  }

  _executeActions(actions) {
    actions.forEach((action) => {
      if (OPERATORS.has(action)) {
        this._performArithmetic(action);
      } else if (STACK_COMMANDS.has(action)) {
        this._manipulateStack(action);
      } else if (action.match(/^-?\d+$/)) {
        this._stack.push(Number(action));
      } else {
        throw Error("Unknown command");
      }
    });
  }

  evaluate(rawInput) {
    const inputTerms = rawInput.toLowerCase().split(" ");

    if (inputTerms[0] === ":") {
      this._defineCommand(inputTerms);
      return;
    }

    this._executeActions(this._expandActions(inputTerms));
  }

  get stack() {
    return this._stack;
  }
}